source/extensions/filters/http/grpc_json_reverse_transcoder/filter_config.h

```diff
@@ -0,0 +1,98 @@
+// translators are reversed.
+using RequestTranslator = ::google::grpc::transcoding::ResponseToJsonTranslator;
+using ResponseTranslator = ::google::grpc::transcoding::JsonRequestTranslator;
+using ::google::grpc::transcoding::MessageStream;
+using ::google::grpc::transcoding::Transcoder;
+using ::google::grpc::transcoding::TranscoderInputStream;
+using ::google::grpc::transcoding::TypeHelper;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace GrpcJsonReverseTranscoder {
+
+// TranscoderImpl wraps the request and response translators and provides access
+// the input and output streams for the translators.
+class TranscoderImpl : public Transcoder {
+public:
+  explicit TranscoderImpl(std::unique_ptr<RequestTranslator> request_translator,
+                          std::unique_ptr<ResponseTranslator> response_translator)
+      : request_translator_(std::move(request_translator)),
+        response_translator_(std::move(response_translator)),
+        response_message_stream_(response_translator_->Output()),
+        request_stream_(request_translator_->CreateInputStream()),
+        response_stream_(response_message_stream_.CreateInputStream()) {}
+
+  TranscoderInputStream* RequestOutput() override { return request_stream_.get(); }
+
+  absl::Status RequestStatus() override { return request_translator_->Status(); }
+
+  Protobuf::io::ZeroCopyInputStream* ResponseOutput() override { return response_stream_.get(); }
+
+  absl::Status ResponseStatus() override { return response_message_stream_.Status(); }
+
+private:
+  std::unique_ptr<RequestTranslator> request_translator_;
+  std::unique_ptr<ResponseTranslator> response_translator_;
+  MessageStream& response_message_stream_;
+  std::unique_ptr<TranscoderInputStream> request_stream_;
+  std::unique_ptr<TranscoderInputStream> response_stream_;
+};
+
+class GrpcJsonReverseTranscoderConfig : public Router::RouteSpecificFilterConfig,
+                                        public Envoy::Logger::Loggable<Envoy::Logger::Id::config> {
+public:
+  GrpcJsonReverseTranscoderConfig(
+      const envoy::extensions::filters::http::grpc_json_reverse_transcoder::v3::
+          GrpcJsonReverseTranscoder& transcoder_config,
+      Api::Api& api);
+
+  // Takes the value of the path header of a gRPC request and returns its path descriptor.
+  const Protobuf::MethodDescriptor* GetMethodDescriptor(absl::string_view path) const;
+
+  // Checks if the request body field is of type `google.api.HttpBody`.
+  bool IsRequestNestedHttpBody(const Protobuf::MethodDescriptor* method_descriptor,
+                               const std::string& request_body_field) const;
+
+  absl::StatusOr<std::unique_ptr<Transcoder>>
+  CreateTranscoder(const Protobuf::MethodDescriptor* method_descriptor,
+                   TranscoderInputStream& request_input,
+                   TranscoderInputStream& response_input) const;
+
+  absl::optional<uint32_t> max_request_body_size_;
```

Comment:
I strongly recommend making all members of the config object const, since it is shared across multiple threads. This will help to be ensure that it is immutable and hence thread safe.

Reply:
Changed the type of `config_` in filter class from `std::shared_ptr<GrpcJsonReverseTranscoderConfig>` to  `std::shared_ptr<cosnt GrpcJsonReverseTranscoderConfig>`, this should make it immutable.